lucet-runtime/lucet-runtime-internals/src/instance.rs

```diff
@@ -800,41 +800,58 @@ impl Instance {
+        // * yielded: state should be `Yielded`; take the yielded value and return a RunResult
+        // * trapped: state should be `Fault`; populate details and return an error or call a hanlder as appropriate
+        // * terminated: state should be `Terminated`; take the termination details and return as an Err
         // * function body returned: set state back to `Ready` with return value
+        //
+        // The state should never be `Ready` at this point
 
-        match &self.state {
+        match &mut self.state {
             State::Running => {
                 let retval = self.ctx.get_untyped_retval();
                 self.state = State::Ready { retval };
                 Ok(RunResult::Returned(retval))
             }
-            State::Terminated { details, .. } => Err(Error::RuntimeTerminated(details.clone())),
-            State::Yielded { val, .. } => Ok(RunResult::Yielded(val.clone())),
-            State::Fault { .. } => {
+            State::Terminated { details, .. } => {
+                Err(Error::RuntimeTerminated(details.take().ok_or_else(
+                    || lucet_format_err!("terminated state contained no details"),
+                )?))
+            }
+            State::Yielded { val, .. } => {
+                Ok(RunResult::Yielded(val.take().ok_or_else(|| {
+                    lucet_format_err!("yielded state contained no value")
+                })?))
+            }
+            State::Fault {
+                ref mut details, ..
+            } => {
                 // Sandbox is no longer runnable. It's unsafe to determine all error details in the signal
                 // handler, so we fill in extra details here.
-                self.populate_fault_detail()?;
-                if let State::Fault { ref details, .. } = self.state {
-                    if details.fatal {
-                        // Some errors indicate that the guest is not functioning correctly or that
-                        // the loaded code violated some assumption, so bail out via the fatal
-                        // handler.
-
-                        // Run the C-style fatal handler, if it exists.
-                        self.c_fatal_handler
-                            .map(|h| unsafe { h(self as *mut Instance) });
-
-                        // If there is no C-style fatal handler, or if it (erroneously) returns,
-                        // call the Rust handler that we know will not return
-                        (self.fatal_handler)(self)
-                    } else {
-                        // leave the full fault details in the instance state, and return the
-                        // higher-level info to the user
-                        Err(Error::RuntimeFault(details.clone()))
-                    }
+                //
+                // FIXME after lucet-module is complete it should be possible to fill this in without
```

Comment:
since the line is highlighted: lucet-module should have all the information for this, it's just a `lucet-runtime-internals` tweak at this point - `lookup_trapcode` looks up faulting function from `rip` purely from module data, but the `addr_details` replacement would just want the name of the faulting function. I dunno if we have the guest module's file name though?

I see this is just a comment that moved up a bit, so this isn't to detract from the changes at all

Reply:
We should totally make that change once we have the guest filename